src/job_queue/queue.rs

```diff
@@ -307,30 +124,228 @@ impl<P: Ord + Send + Sync + 'static> JobQueue<P> {
+// are as small as possible.
+//
+// stop:
+//
+// When a 'stop' msg is received we send a cancel msg to the current job (if any) and
+// wait for it to complete. Then we exit the loop and return.
+async fn process_jobs<P: Ord + Send + Sync + 'static>(
+    shared_queue: Arc<Mutex<SharedQueue<P>>>,
+    mut rx_stop: watch::Receiver<()>,
+    mut rx_job_added: mpsc::UnboundedReceiver<()>,
+) {
+    // job number starts at 1 and increments with each job that is processed.
+    // note that processing order may be different than order in which jobs
+    // are added due to job priorities.
+    let mut job_num: usize = 1;
+
+    // loop until 'stop' msg is received or job_added channel is closed.
+    //
+    // note:  this unbounded channel will grow in size as new job(s) are
+    // added while an existing job is running.  ie, we read from the
+    // channel after each job completes.
+    while rx_job_added.recv().await.is_some() {
+        // Find the next job to run, and the number of jobs left in queue
+        tracing::debug!("task process_jobs received JobAdded message.");
+        let (next_job, num_pending) = {
+            // acquire mutex lock
+            let mut guard = shared_queue.lock().unwrap();
+
+            // pick the highest priority job
+            guard
+                .jobs
+                .make_contiguous()
+                .sort_by(|a, b| b.priority.cmp(&a.priority));
+            let job = guard.jobs.pop_front().unwrap();
+
+            // set highest priority job as the current job
+            guard.current_job = Some(CurrentJob {
+                job_num,
+                job_id: job.job_id,
+                cancel_tx: job.cancel_tx.clone(),
+            });
+
+            (job, guard.jobs.len())
+        }; // mutex lock is released when guard drops.
+
+        // log that we are starting a job
+        tracing::info!(
+            "  *** JobQueue: begin job #{} - {} - {} queued job(s) ***",
+            job_num,
+            next_job.job_id,
+            num_pending
+        );
+
+        // record time that job starts
+        let timer = tokio::time::Instant::now();
+
+        // spawn task that performs the job, either async or blocking.
+        let task_handle = if next_job.job.is_async() {
+            tokio::spawn(
+                async move { next_job.job.run_async_cancellable(next_job.cancel_rx).await },
+            )
+        } else {
+            tokio::task::spawn_blocking(move || next_job.job.run(next_job.cancel_rx))
+        };
+
+        let task_result = tokio::select! {
+            // handle msg over 'stop' channel which indicates we must exit the loop.
+            _ = rx_stop.changed() => {
+                tracing::debug!("task process_jobs received Stop message.");
+
+                // acquire mutex lock and obtain current_job info, if any.
+                let maybe_info = shared_queue.lock().unwrap().current_job.as_ref().map(|cj| (cj.job_id, cj.cancel_tx.clone()) );
```

Comment:
Formatting issues. Unfortunately, `rustfmt` doesn't work inside the `select!` macro. :confused: (See also: https://github.com/tokio-rs/tokio/issues/5441.)

Reply:
ah.  today i learned.  thx.

Reply:
move stop msg handling into its own fn, which is formatted by rustfmt, in 7fc22f78b47a08ffe0899442cb35affdf2babb67